nexus/pack.cc

```diff
@@ -2075,6 +2075,239 @@ struct NexusPacker
+            return nullptr;
+        }
+
+        // Get clock nets of IOLOGIC and the flip-flop
+        if (iol != nullptr) {
+            NetInfo* iol_c = get_net_or_empty(iol, id_SCLKOUT);
+            NetInfo* ff_c  = get_net_or_empty(ff,  id_CLK);
+
+            // If one of them is floating or it is not the same net then abort
+            if (iol_c == nullptr || ff_c == nullptr) {
+                return nullptr;
+            }
+            if (iol_c->name != ff_c->name) {
+                return nullptr;
+            }
+        }
+
+        // Get reset nets of IOLOGIC and the flip-flop
+        if (iol != nullptr) {
+            NetInfo* iol_r = get_net_or_empty(iol, id_LSROUT);
+            NetInfo* ff_r  = get_net_or_empty(ff,  id_LSR);
+
+            // If one of them is floating or it is not the same net then abort.
+            // But both can be floating.
+            if (!(iol_r == nullptr && ff_r == nullptr)) {
+                if (iol_r == nullptr || ff_r == nullptr) {
+                    return nullptr;
+                }
+                if (iol_r->name != ff_r->name) {
+                    return nullptr;
+                }
+            }
+        }
+
+        // FIXME: Check if the flip-flop has:
+        // - non-inverted clock
+        // - same reset "type" as ODDR
+        // - others ?
+
+        return ff;
+    }
+
+    // IOLOGIC requires some special handling around itself and IOB. This
+    // function does that.
+    void handle_iologic() {
+
+        // Map of flip-flop cells that drive IOLOGIC+IOB pairs
+        dict<IdString, std::vector<std::pair<IdString,IdString>>> tff_map;
+
+        for (auto &cell : ctx->cells) {
+            CellInfo* iol = cell.second.get();
+            if (iol->type != id_SIOLOGIC && iol->type != id_IOLOGIC) {
+                continue;
+            }
+
+            bool isIDDR = false;
+            bool isODDR = false;
+
+            CellInfo *iob = nullptr;
+            NetInfo *di = get_net_or_empty(iol, id_DI);
+            if (di != nullptr && di->driver.cell != nullptr) {
+                iob = di->driver.cell;
+                isIDDR = true;
+            }
+            NetInfo *dout = get_net_or_empty(iol, id_DOUT);
+            if (dout != nullptr && dout->users.size() == 1) {
+                iob = dout->users.at(0).cell;
+                isODDR = true;
+            }
+            NetInfo *tout = get_net_or_empty(iol, id_TOUT);
+            if (tout != nullptr && tout->users.size() == 1) {
+                iob = tout->users.at(0).cell;
+                isODDR = true; // FIXME: Not sure
+            }
+            NPNR_ASSERT(iob != nullptr);
+
+            // SIOLOGIC handling
+            if (iol->type == id_SIOLOGIC) {
+
+                // We have IDDR+ODDR
+                if (isODDR && isIDDR) {
+                    if (!iob->attrs.count(ctx->id("GLITCHFILTER"))) {
+                        iob->attrs[ctx->id("GLITCHFILTER")] = std::string("ON");
```

Comment:
to double check, are you sure this defaults to `ON` in the vendor tools?

Reply:
So I did a minitest considering various `IDDRX1` and `ODDRX1` configurations:
 - regular bidirectional IO
 - IDDRX1 only
 - ODDRX1 only
 - ODDRX1 plus tri-state control (external to the ODDR)
 - IDDRX1 + ODDRX1 plus external tri-state control
 
And I noticed the glitch filter enabled only for the bidirectional DDR case by default. Those attributes assigned around this place in the code are the result of my findings.

Reply:
Great, would you be willing to submit a PR with these tests into https://github.com/gatecat/prjoxide/tree/master/minitests for future reference?

Reply:
Sure. I just need to clean them a little bit.